### PR TITLE
feat: added dataset recall testing API

### LINKED ISSUE #8959: search on database through api
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

I'm looking for a rag repository platform that supports API calls, and I found this feature available in dify's congeneric product fastgpt (https://doc.fastgpt.cn/docs/development/openapi/dataset/#%E6%90%9C%E7%B4%A2%E6%B5%8B%E8%AF%95).
I didn't find this feature in dify's API documentation (https://docs.dify.ai/guides/knowledge-base/maintain-dataset-via-api).

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by gubinjie**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

A knowledge base recall test interface for external calls is added

Fixes https://github.com/langgenius/dify/issues/8959
Close https://github.com/langgenius/dify/pull/8477

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: web/app/(commonLayout)/datasets/template/template.en.mdx [hunk 1]
```diff
@@ -1050,6 +1050,151 @@ import { Row, Col, Properties, Property, Heading, SubProperty, Paragraph } from
...
+          Unused field
+      </Property>
+    </Properties>
+  </Col>
+  <Col sticky>
+    <CodeGroup
+      title="Request"
+      tag="GET"
+      label="/datasets/{dataset_id}/hit_testing"
+      targetCode={`curl --location --request GET '${props.apiBaseUrl}/datasets/{dataset_id}/hit_testing' \\\n--header 'Authorization: Bearer {api_key}'\\\n--header 'Content-Type: application/json'\\\n--data-raw '{
```
Comment (hwzhuhao) [inline web/app/(commonLayout)/datasets/template/template.en.mdx:1099]: This URL should be datasets2/{dataset_id}/hit_testing, right?

File: api/controllers/service_api/dataset/hit_testing.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from core.model_runtime.errors.invoke import InvokeError
+from fields.hit_testing_fields import hit_testing_record_fields
+from services.dataset_service import DatasetService
+from services.hit_testing_service import HitTestingService
+
+
+class HitTestingApi(DatasetApiResource):
+    def post(self, tenant_id, dataset_id):
+        dataset_id_str = str(dataset_id)
+        logging.error(f"{dataset_id_str}")
```
Comment (hwzhuhao) [inline api/controllers/service_api/dataset/hit_testing.py]: This is debug print information, right? It can be removed.

Comment (hwzhuhao) [reply api/controllers/service_api/dataset/hit_testing.py]: If it's really needed, change the log level to info.

**Comment by hwzhuhao** [comment]
1.Many contents in controllers/service_api/dataset/hit_testing.py and controllers/console/datasets/hit_testing.py are duplicated. Can a common method be abstracted?
2.It would be even better if there were a test screenshot included

**Comment by gubinjie** [comment]
@hwzhuhao 
sorry, I reviewed these two pieces of code and I think it's already an atomic operation that doesn't need to extract generic methods.
The following is a screenshot of the test call interface.
image

**Comment by hwzhuhao** [comment]
@gubinjie Maybe you can add a base class, for example:

class HitTestingBase:
    def get_and_validate_dataset(self, dataset_id_str):
        dataset = DatasetService.get_dataset(dataset_id_str)
        if dataset is None:
            raise NotFound("Dataset not found.")
        try:
            DatasetService.check_dataset_permission(dataset, current_user)
        except services.errors.account.NoPermissionError as e:
            raise Forbidden(str(e))
        return dataset

    def parse_args(self):
        parser = reqparse.RequestParser()
        parser.add_argument("query", type=str, location="json")
        parser.add_argument("retrieval_model", type=dict, required=False, location="json")
        parser.add_argument("external_retrieval_model", type=dict, required=False, location="json")
        return parser.parse_args()
    
    def hit_testing_args_check(self, args):
        HitTestingService.hit_testing_args_check(args)

    def perform_hit_testing(self, dataset, args):
        try:
            response = HitTestingService.retrieve(
                dataset=dataset,
                query=args["query"],
                account=current_user,
                retrieval_model=args["retrieval_model"],
                external_retrieval_model=args["external_retrieval_model"],
                limit=10,
            )
            return {"query": response["query"], "records": marshal(response["records"], hit_testing_record_fields)}
        except services.errors.index.IndexNotInitializedError:
            raise DatasetNotInitializedError()
        except ProviderTokenNotInitError as ex:
            raise ProviderNotInitializeError(ex.description)
        except QuotaExceededError:
            raise ProviderQuotaExceededError()
        except ModelCurrentlyNotSupportError:
            raise ProviderModelCurrentlyNotSupportError()
        except LLMBadRequestError:
            raise ProviderNotInitializeError(
                "No Embedding Model or Reranking Model available. Please configure a valid provider "
                "in the Settings -> Model Provider."
            )
        except InvokeError as e:
            raise CompletionRequestError(e.description)
        except ValueError as e:
            raise ValueError(str(e))
        except Exception as e:
            logging.exception("Hit testing failed.")
            raise InternalServerError(str(e))

class HitTestingApi(Resource, HitTestingBase):
    @setup_required
    @login_required
    @account_initialization_required
    def post(self, dataset_id):
        self.dataset_id_str = str(dataset_id)
        dataset = self.get_and_validate_dataset(self.dataset_id_str)
        args = self.parse_args()
        self.hit_testing_args_check(args)
        return self.perform_hit_testing(dataset, args)

**Comment by gubinjie** [comment]
Okay, big guy, wait a minute, modify it immediately~~0.0

**Comment by gubinjie** [comment]
@hwzhuhao I'm okay, please take a look

**Comment by hwzhuhao** [comment]
@gubinjie The file name might be better as hit_testing_base.py. Additionally, please review the comments and address them. Thank you.
image

**Comment by gubinjie** [comment]
@hwzhuhao Hmm, I found out that the path in the code was wrong before, and now it's been modified

**Review by hwzhuhao — APPROVED** [review]

**Comment by gubinjie** [comment]
These have been corrected

**Comment by hwzhuhao** [comment]
lgtm, thanks. @crazywoola

**Review by crazywoola — APPROVED** [review]

**Pull request merged**